src/Shared/UnitTests/TestEnvironment.cs

```diff
@@ -0,0 +1,258 @@
+        }
+    }
+
+    public static class VariantFactory
+    {
+        /// <summary>
+        ///     Creates a test variant that corresponds to a temporary file under a specific temporary folder. File will
+        ///     be cleaned up when the test completes.
+        /// </summary>
+        /// <param name="folderVariant"></param>
+        /// <param name="extension">Extensions of the file (defaults to '.tmp')</param>
+        public static FileVariant CreateFile(FolderVariant folderVariant, string extension = ".tmp")
+        {
+            return new FileVariant(folderVariant.FolderPath, extension);
+        }
+
+        /// <summary>
+        ///     Creates a test variant that corresponds to a temporary file which will be deleted when the test completes.
+        /// </summary>
+        /// <param name="extension">Extensions of the file (defaults to '.tmp')</param>
+        public static FileVariant CreateFile(string extension = ".tmp")
+        {
+            return new FileVariant(extension);
+        }
+
+        /// <summary>
+        ///     Creates a test variant used to add a unique temporary folder during a test. Will be deleted when the test
+        ///     completes.
+        /// </summary>
+        public static FolderVariant CreateFolder()
+        {
+            return new FolderVariant();
+        }
+
+        /// <summary>
+        ///     Create an test variant used to change the value of an environment variable during a test. Original value
+        ///     will be restored when complete.
+        /// </summary>
+        public static EnvironmentVariant CreateEnvironment(string environmentVariableName, string newValue)
+        {
+            return new EnvironmentVariant(environmentVariableName, newValue);
+        }
+    }
+
+    /// <summary>
+    ///     Things that are expected not to change and should be asserted before and after running.
+    /// </summary>
+    public abstract class TestInvariant
+    {
+        public abstract void AssertInvariant();
+    }
+
+    /// <summary>
+    ///     Things that are expected to change and should be reverted after running.
+    /// </summary>
+    public abstract class TestVariant
+    {
+        public abstract void Revert();
+
+        public FileVariant Create()
+        {
+            return new FileVariant("test");
+        }
+    }
+
+    public class GenericInvariant : TestInvariant
+    {
+        private readonly Func<string> _accessorFunc;
+        private readonly string _name;
+        private readonly string _originalValue;
+
+        public GenericInvariant(string name, Func<string> accessorFunc)
+        {
+            _name = name;
+            _accessorFunc = accessorFunc;
+            _originalValue = accessorFunc();
+        }
+
+        public override void AssertInvariant()
+        {
+            Assert.Equal($"{_name}: {_originalValue}", $"{_name}: {_accessorFunc()}");
+        }
+    }
+
+    public class FileVariant : TestVariant
```

Comment:
I think I'd prefer a name like `TransientTestFile`, and the superclass something like `TransientTestState`. "Variant" here keeps confusing me.